### PR TITLE
ReactChildren.map: only add slash if new child has key

### PR DESCRIPTION
**Pull request by ianobermiller**
See the new test for the scenario I am trying to fix; if you clone an
element in React.cloneElement, vs just returning it directly, you will
get a different key (with a slash in front) even though the two
children are identical.

Test Plan:

`npm test -- src/isomorphic/children/__tests__/ReactChildren-test.js`

### CODE REVIEW

**Comment by ianobermiller** [comment]
cc @spicyj

File: src/isomorphic/children/ReactChildren.js [hunk 1]
```diff
@@ -117,7 +117,7 @@ function mapSingleChildIntoContext(bookKeeping, child, childKey) {
         // traverseAllChildren used to do for objects as children
         keyPrefix +
         (
-          mappedChild !== child ?
+          mappedChild !== child && mappedChild.key ?
           escapeUserProvidedKey(mappedChild.key || '') + '/' :
```
Comment (zpao) [inline src/isomorphic/children/ReactChildren.js]: Assume Ben likes this, can you drop the `|| ''` here since we already know we have it.

Comment (spicyj) [inline src/isomorphic/children/ReactChildren.js]: Haven't thought about this in super detail yet but why not `mappedChild.key !== child.key` as I suggested in chat?

**Comment by ianobermiller** [comment]
Nice catch, thanks.

File: src/isomorphic/children/ReactChildren.js (hunk 1 above)
Comment (ianobermiller) [inline src/isomorphic/children/ReactChildren.js]: Yes, that make sense, need to add a null-check then. Still need a truthy check on `mappedChild.key`, so `mappedChild.key && (!child || mappedChild.key !== child.key)`

File: src/isomorphic/children/ReactChildren.js [hunk 2]
```diff
@@ -117,8 +117,8 @@ function mapSingleChildIntoContext(bookKeeping, child, childKey) {
         // traverseAllChildren used to do for objects as children
         keyPrefix +
         (
-          mappedChild !== child ?
-          escapeUserProvidedKey(mappedChild.key || '') + '/' :
+          (mappedChild.key && (!child || (child.key !== mappedChild.key))) ?
```
Comment (ianobermiller) [inline src/isomorphic/children/ReactChildren.js]: I hate all the parens, but this makes it crystal clear.

**Comment by RoccoC** [comment]
Curious if/when this PR will be merged into master? This PR will fix a few problems we are dealing with in our project. Thanks!

**Comment by Erid** [comment]
I'm unfamiliar with the process... When can we expect it to be published on npm?

**Comment by jimfb** [comment]
After the next release major release, which is probably less than a month away :).

**Pull request merged**